pkg/apis/apis/v1alpha1/types_apiresourceschema.go

```diff
@@ -163,3 +163,88 @@ func (v *APIResourceVersion) SetSchema(schema *apiextensionsv1.JSONSchemaProps)
 	v.Schema.Raw = raw
 	return nil
 }
+
+// +crd
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Cluster,categories=kcp
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+
+// APIConversion contains rules to convert between different API versions in an APIResourceSchema. The name must match
+// the name of the APIResourceSchema for the conversions to take effect.
+type APIConversion struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Spec holds the desired state.
+	Spec APIConversionSpec `json:"spec"`
+}
+
+// APIConversionSpec contains rules to convert between different API versions in an APIResourceSchema.
+type APIConversionSpec struct {
+	// +required
+	// +listType=map
+	// +listMapKey=from
+	// +listMapKey=to
+
+	// conversions specify rules to convert between different API versions in an APIResourceSchema.
+	Conversions []APIVersionConversion `json:"conversions"`
+}
+
+// APIVersionConversion contains rules to convert between two specific API versions in an
+// APIResourceSchema. Additionally, to avoid data loss when round-tripping from a version that
+// contains a new field to one that doesn't and back again, you can specify a list of fields to
+// preserve (these are stored in annotations).
+type APIVersionConversion struct {
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=^v[1-9][0-9]*([a-z]+[1-9][0-9]*)?$
+
+	// from is the source version.
+	From string `json:"from"`
+
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=^v[1-9][0-9]*([a-z]+[1-9][0-9]*)?$
+
+	// to is the target version.
+	To string `json:"to"`
+
+	// +required
+	// +listType=map
+	// +listMapKey=to
+
+	// rules contains field-specific conversion expressions.
+	Rules []APIConversionRule `json:"rules"`
+
+	// +optional
+
+	// preserve contains a list of paths to fields to preserve, relative to the root of the object, such as
+	// 'spec.name.first'.
+	Preserve []string `json:"preserve,omitempty"`
+}
+
+// APIConversionRule specifies how to convert a single field.
+type APIConversionRule struct {
```

Comment:
how can I delete fields?

Reply:
nvmd. Deletion is not an issue with the compute model.

But wondering: what is the new object before the first rule? Empty? Or equal where types match?

Reply:
And if the later, what is the type equality applied here?

Reply:
It's up to us. Currently, `to` starts as a complete copy of `from`. I believe because unknown fields are not preserved, any places where the types don't match, those fields are dropped.

Reply:
ic, so basically a unstructured copy that is then pruned.

> any places where the types don't match

This we have to understand very well. What does it mean to not match? Is it basically JSON marshal-unmarshal compatibility?

Reply:
I mistyped. Any places where the field paths don't match, those fields are dropped. e.g. if v1 has spec.firstName and v2 has spec.name.first, the conversion will create an map that looks like this

```yaml
spec:
  firstName: Andy
  name:
    first: Andy
```

and then spec.firstName is pruned (somewhere - haven't tracked down the specific place) so by the time you get it out as v2, the only thing left is spec.name.first